source/internal/_clone.js

```diff
@@ -12,29 +11,104 @@ import type from '../type';
  * @param {Boolean} deep Whether or not to perform deep cloning.
  * @return {*} The copied value.
  */
-export default function _clone(value, refFrom, refTo, deep) {
+export default function _clone(value, deep, map) {
+
+  map || (map = new _ObjectMap());
+
+  // this avoids the slower switch with a quick if decision removing some milliseconds in each run.
+  if (_isPrimitive(value)) {
+    return value;
+  }
+
   var copy = function copy(copiedValue) {
-    var len = refFrom.length;
-    var idx = 0;
-    while (idx < len) {
-      if (value === refFrom[idx]) {
-        return refTo[idx];
-      }
-      idx += 1;
+    // Check for circular and same references on the object graph and return its corresponding clone.
+    var cachedCopy = map.get(value);
+
+    if (cachedCopy) {
+      return cachedCopy;
     }
-    refFrom[idx] = value;
-    refTo[idx] = copiedValue;
+
+    map.set(value, copiedValue);
+
     for (var key in value) {
-      copiedValue[key] = deep ?
-        _clone(value[key], refFrom, refTo, true) : value[key];
+      copiedValue[key] = deep ? _clone(value[key], true, map) : value[key];
     }
     return copiedValue;
   };
+
   switch (type(value)) {
-    case 'Object':  return copy({});
-    case 'Array':   return copy([]);
-    case 'Date':    return new Date(value.valueOf());
-    case 'RegExp':  return _cloneRegExp(value);
-    default:        return value;
+    case 'Object':
+      return copy({});
+    case 'Array':
+      return copy([]);
+    case 'Date':
+      return new Date(value.valueOf());
+    case 'RegExp':
+      return _cloneRegExp(value);
+    default:
+      return value;
   }
 }
+
+function _isPrimitive(param) {
+  var type = typeof param;
+  return param == null || (type != 'object' && type != 'function');
+}
+
+function _ObjectMap() {
+  this.map = {};
+  this.length = 0;
+}
+
+_ObjectMap.prototype.set = function(key, value) {
+  const hashedKey = this.hash(key);
+
+  let bucket = this.map[hashedKey];
+  if (!bucket) {
+    this.map[hashedKey] = bucket = [];
+  }
+
+  bucket.push([key, value]);
+  this.length += 1;
+};
+
+_ObjectMap.prototype.hash = function(key) {
+  let hashedKey = [];
+  for (var value in key) {
+    hashedKey.push(key[value]);
+  }
+  return hashedKey.join();
+};
```

Comment:
I think there's a problem with any generic hash here.  For this simple one, note that 

```js
hash ({foo: 1, bar: 2}) == hash ({bar: 1, foo: 2}) //=> true
hash ({foo: {a: 42}}) == hash ({foo: {b: false}})  //=> true
```

That second one is especially problematic.  We could easily end up with data that's all hash collisions.

But even a more sophisticated hash can still have such problems.  I suppose if we get sophisticated enough, then we'd probably lose much of the benefit.

Reply:
Yes, the hash function is very simple and if we made it more sophisticated we end up losing performant, but I think this hash function is just good enough to split the data a little bit to make it performant, and yes we could in a worst-case scenario end up with a collection full of collisions, but in that case, the get function will just behave the way the old algorithm behaved.

Reply:
I'm not convinced.

I'd love to see some benchmarks with different styles of input objects.

This hash seems too tuned to that input structure.

Reply:
After further inspection, I realize I made a mistake on my hash function, instead of using just the value I should use the propertyName + value as the hash, but regardless, the idea still the same.
About the benchmarks with different inputs I used the following array:
````js
const obja = { foo: "foo1", bar: "bar1" };
const objb = { foo: "foo2", bar: "bar2" };
const objc = { foo: "foo3", bar: "bar3" };
const objd = { foo: "foo4", bar: "bar4" };


const arr = [
  { obj1: obja, obj2: { foo: "foo1", bar: "bar1" }, obj3: objc, obj4: { foo: "foo2", bar: "bar2" }, obj5: obja, obj6: objb },
  { obj1: obja, obj2: { foo: "foo1", bar: "bar1" }, obj3: objd, obj4: { foo: "foo2", bar: "bar2" }, obj5: obja, obj6: objb },
  { obj1: obja, obj2: { foo: "foo1", bar: "bar1" }, obj3: objc, obj4: { foo: "foo2", bar: "bar2" }, obj5: obja, obj6: objb },
  { obj1: obja, obj2: { foo: "foo1", bar: "bar1" }, obj3: objd, obj4: { foo: "foo2", bar: "bar2" }, obj5: obja, obj6: objb },
];
````

and got the following results:

max/min/avg time in ms after 100 runs
ramda: new clone:             max: 0.815 min: 0.030 avg: 0.059 
ramda: old clone:             max: 0.816 min: 0.029 avg: 0.063 
lodash cloneDeep():           max: 1.913 min: 0.038 avg: 0.114 
JSON.parse(JSON.stringify()): max: 0.804 min: 0.013 avg: 0.030 

Because it is a small object to be cloned it is the best-case scenario for the old clone() but still, the new one shows better results, it only gets better as the object size goes up.
I am happy to test other examples of objects or even show the code that I'm using to benchmark so the numbers could be confirmed.